### ISSUE TITLE
Pyright 373 regression - code unreachable 

### ISSUE THREAD

**Issue by danielpopescu** [microsoft/pyright#8518]
Code Sample

```python
from __future__ import annotations

from typing import Callable, Iterable, Iterator


def fold_left[T, U](lst: Iterable[T], init: U, f: Callable[[U, T], U]) -> U:
    def fold_left_it(it: Iterator[T], init: U, f: Callable[[U, T], U]) -> U:
        elem: T | None = next(it, None)
        return fold_left_it(it, f(init, elem), f) if elem is not None else init

    if isinstance(lst, Iterator):
        it: Iterator[T] = lst
        return fold_left_it(it, init, f)
    return fold_left_it(iter(lst), init, f)
```

Latest version shows that  the lines

```
        it: Iterator[T] = lst
        return fold_left_it(it, init, f)

```
are unreachable, presumably due to always failing isinstance call.

Python 3.12, pyright 373. 

Thank you

**Comment by erictraut** [comment]
Thanks for the bug report. This will be addressed in the next release.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.374.
